### ISSUE TITLE
Let Stage plugins advertise extensions they answer

### ISSUE THREAD

**Issue by hobu** [PDAL/PDAL#2714]
Plugin developers still have to hack StageExtensions to get PDAL to answer their plugins. Plugins should be able to advertise the extensions they answer at runtime.

* What to do about multiple drivers answering the same extension? I say last one registered wins for now.

* What to do about prefixing like `readers.ept` does?

**Comment by abellgithub** [comment]
This doesn't work. How would you get the extension without loading the
plugin?

**Comment by hobu** [comment]
> This doesn't work. 

You mean this doesn't work *right now* 😄 For external developers using PDAL and writing their own plugins, our API means they must make a pull request to the main repository to get the Stage machinery to answer their plugin **or** they must fork their own PDAL and hack in the support that they need. Both of these alternatives are bad.

**Comment by abellgithub** [comment]
No, I mean this won't work.  As I said, you can't determine anything from the plugin without opening it unless you're going to do something bad like put the information in the filename.

I had this done and working a couple of releases ago by sticking the information in a JSON file that would sit next to the PDAL executable (or some such -- I don't remember exactly), but you didn't like it, so it got killed.  I pulled the remainder of that code out just prior to 2.0 release.

**Comment by hobu** [comment]
> but you didn't like it, so it got killed.

Let's reconsider this. I didn't want to have to add a `PDAL_DATA` or some such directory. Presumably users would want a user-overrideable local file in addition to the system-installed one.

**Comment by runette** [comment]
Why not use the existing stage invocation functions to do this?

I mean :

- Have the equivalent of readers.custom that takes a parameter of the filename of the reader executable - i.e. the equivalent of 

```
{
     "type": "readers.custom",
     "executable": "myReader.dll"
}
```
- The first thing it does is call a registration method in `myReader.dll` to make sure that it is a valid PDAL reader and to get the necessary details. If it fails at this stage - throw the input file into the normal filename based reader selection process, and
- provide a mechanism (I have not created a PDAL stage so I don't know if it already exists) for `myReader.dll` to return the input file as "not my concern" in which case the input file is thrown into the normal selection process. This answers the scenario in ["Allowing selection of PDAL shared readers by file extension?"](https://lists.osgeo.org/pipermail/pdal/2020-September/002258.html).

This all at least has the benefit of avoiding any new configuration files etc and avoids opening up holes for potential spoofing of readers (I am not sure why anyone would want to spoof a reader but always good to avoid creating the possibility).

Just thinking aloud ...

Paul

**Comment by abellgithub** [comment]
Hi Paul,

I don't think your idea exactly addresses this problem or that of the linked mailing list question. The issue is generally not one of pipelines, since specifying "type" always maps to a particular stage DLL file. I suppose one might want to use a different filename for a stage than the one required by that mapping, but this issue and the mailing list question seem to me about mapping filenames used in PDAL utilities on the command line to stages. In your example, you already know the stage that you want invoked.

Maybe I'm missing something, though.

**Comment by styrso** [comment]
Hello,

I posted the question to the mailing list. The issue is that if you've a multi-input stage such as the delta kernel, it's not possible to mix custom readers and default readers for the input files.

One possibility might be to 

- add a method to custom readers that allows them to return a list of file extensions that they handle
- have a supplemental readers flag on the command line, e.g. -s custom.reader (allow multiple -s entries)

Any reader(s) specified using -s would be loaded and have their file handling capabilities queried. These capabilities would be added to the list of file types handled already by the default PDAL io plugins, and the input files then parsed using the appropriate reader from the default and custom readers.

Does that make any sense?

**Comment by longhuan2018** [comment]
Why don't we let CREATE_SHARED_STAGE switch from PluginInfo to StaticPluginInfo? StaticPluginInfo contains supported extensions，Once the plug-in is loaded, it's like CREATE_STATIC_STAGE behaves the same, It can be loaded by StageFactory::inferReaderDriver() or StageFactory::inferWriterDriver().
What we need to do is to add an PDAL_DLL  for StageExtensions.  #3609

**Comment by abellgithub** [comment]
We want to avoid loading all the plugins just to get the extensions.
